minimal_scheduling/minimal_sched_fifo_callback_group.cpp

```diff
@@ -0,0 +1,151 @@
+// Copyright 2022 Carlos San Vicente
+// Copyright 2016 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <pthread.h>
+
+#include <chrono>
+#include <memory>
+#include <string>
+
+#include "rclcpp/rclcpp.hpp"
+#include "std_msgs/msg/string.hpp"
+
+/* For this example, we will be using multiple callback groups to separate callbacks, so it is
+ * possible to spin them in threads with different scheduling priorities. */
+
+using namespace std::chrono_literals;
+
+class MinimalPublisher : public rclcpp::Node
+{
+public:
+  MinimalPublisher()
+  : Node("minimal_publisher"), count1_(0), count2_(0)
+  {
+    publisher1_ = this->create_publisher<std_msgs::msg::String>("topic", 10);
+    auto timer_callback1 =
+      [this]() -> void {
+        auto message = std_msgs::msg::String();
+        message.data = "Hello, world! " + std::to_string(this->count1_++);
+        RCLCPP_INFO(this->get_logger(), "Publishing     : '%s'", message.data.c_str());
+        this->publisher1_->publish(message);
+      };
+    timer1_ = this->create_wall_timer(500ms, timer_callback1);
+
+    realtime_callback_group_ = this->create_callback_group(
+      rclcpp::CallbackGroupType::MutuallyExclusive, false);
+
+    publisher2_ = this->create_publisher<std_msgs::msg::String>("topic_rt", 10);
+    auto timer_callback2 =
+      [this]() -> void {
+        auto message = std_msgs::msg::String();
+        message.data = "Hello, world! " + std::to_string(this->count2_++);
+        RCLCPP_INFO(this->get_logger(), "Publishing (RT): '%s'", message.data.c_str());
+        this->publisher2_->publish(message);
+      };
+    timer2_ = this->create_wall_timer(500ms, timer_callback2, realtime_callback_group_);
```

Comment:
Question: is the sleep done by ROS via `clock_nanosleep` with CLOCK_MONOTONIC? maybe something to confirm...? :thinking:

Reply:
The wait depends on each RMW implementation but it is expected to use `CLOCK_MONOTONIC`.  See these related issues: 
- https://github.com/ros2/rcl/blob/4740c82864518a331ae98799f25b2ba085b22473/rcl/include/rcl/timer.h#L64-L159
- http://togithub.com//ros2/ros2/issues/1080
- http://togithub.com/ros2/rmw/pull/316
- http://togithub.com//ros2/rmw_cyclonedds/issues/276

I guess your question is whether timers can be used for real-time execution. This is something we can discuss in the next meeting. AFAIK, in some RMW implementations, this ultimately relies on `pthread_cond_wait` which has a mutex and cannot be configured with priority inheritance. See https://wiki.linuxfoundation.org/realtime/documentation/known_limitations.

Reply:
Maybe this is something we can create a thread in the ros-realtime/community issues? It doesn't belong in this PR but the information you have here is good for future reference and discussions.